{% extends 'chado/base.html' %}
{% load static %}
{% block css %}
<link rel="stylesheet" type="text/css" href="{% static 'css/context.css' %}">
<style>

#toggle {
	width: 100%;
	height: 200px;
	background-color: rgb(238, 238, 238);
	position: fixed;
	bottom: 0;
	left: 0;
	box-shadow: 0px 1px 4px rgba(0, 0, 0, 0.2);
	padding: 0 10px;
	text-align: left;
	overflow: auto;
}
#toggle h4 {
	word-wrap:break-word;
}
#viewer, #legend {
	margin: 0;
	padding: 0;
}

#viewer {
	background-color: #f6f6f6;
	border-bottom: 1px solid #eee;
}

#legend {
	position: fixed;
	top: 50px;
	right: 0;
	bottom: 0;
	overflow: auto;
	background: #F6F6F6;
}

.sticky {  
    position: fixed;  
    left: 0;  
    top: 50px; /* below the main navbar */ 
    z-index: 100;
}
</style>
{% endblock %}
{% block content %}
<div class="page-header" id="banner">
    <div class="row">
        <div class="col-lg-12">
            <h1>Synteny Miner</h1>
        </div>
    </div>
</div>
<div class="row">
    <div class="col-lg-12">
        {% if messages %}
            {% for message in messages %}
            <div class="alert{% if message.tags %} alert-{{ message.tags }}{% endif %}">
                <strong>{{ message.tags|capfirst }}!</strong> {{ message }}
            </div>
            {% endfor %}
        {% endif %}
    </div>
</div>
{% endblock %}
{% block bottom %}
<div id="remember">
	<div class="col-lg-9" id="viewer"></div>
</div>
<div class="col-lg-9" id="synteny">
</div>
<div class="col-lg-3" id="legend">
</div>
<div id="toggle"></div>
{% endblock %}
{% block javascript %}
<script type="text/javascript" src="{% static 'js/jquery-1.10.2.min.js' %}"></script>
<script type="text/javascript" src="{% static 'js/jquery-ui-1.10.3.custom.min.js' %}"></script>
<script type="text/javascript" src="http://d3js.org/d3.v3.min.js"></script>
<script type="text/javascript" src="{% static 'js/context.js' %}"></script>
<script type="text/javascript" src="{% static 'js/context_synteny.js' %}"></script>
<script type="text/javascript" src="{% static 'js/context_viewer.js' %}"></script>
<script type="text/javascript" src="{% static 'js/context_legend.js' %}"></script>
<script type="text/javascript" src="{% static 'js/smith.js' %}"></script>
<script type="text/javascript" src="{% static 'js/enhancement.js' %}"></script>
<script type="text/javascript">
	// shared variables
	var json = '{{ json|safe }}',
		data = JSON.parse(json),
        context_data = JSON.parse(json),
		color = d3.scale.category20(),
		family_names = get_family_name_map( data );
	var remove_image = "{% static 'gif/close.png' %}";
	
	$(document).ready(function() {
		// hide the toggle div
        $('#toggle').hide();
		// make the viewer sticky
	    var stickyContextTop = $('#viewer').offset().top;  
	    var stickyContext = function(){  
	        var scrollTop = $(window).scrollTop()+50; // pad for the main navbar
	        if (scrollTop > stickyContextTop) {   
				$("#remember").height($("#viewer").height());
	            $('#viewer').addClass('sticky');  
				} else {
				$("#remember").height($("#viewer").height());
	            $('#viewer').removeClass('sticky');   
	        } 
	    };
	    stickyContext();  
	    $(window).scroll(function() {  
	        stickyContext();  
	    });
	});

	// hide the toggle div when the click isn't in it or on a gene
	$(document).on('click', function(e) {
		var bool = $(e.target).closest('#toggle').length == 0
				&& $(e.target).closest('.axis').length == 0
				&& $(e.target).closest('.gene').length == 0
				&& $(e.target).closest('.legend').length == 0
				&& $("#toggle").is(":visible");
		if( bool ) {
			$('#toggle').toggle('slide', {direction:'down'});
			//$('body').css('margin-bottom', '0px');
		}
	});

	// a function that shows to toggle div
	function show_toggle() {
		// don't toggle it unless it's currently hidden
		if (!$('#toggle').is(':visible')) {
			$('#toggle').toggle('slide', {direction:'down'});
			//$('body').css('margin-bottom', '200px');
		}
	}

	// gene clicked callback function
	function gene_clicked( selection ) {
		$.ajax({
	        url:'/chado/phylo_slide/node/slide/',
	        data: {
	        // this is actually a phylonode_id
	        gene: selection.id,
	        // all pages are protected against csrf so we need to pass the token for this page
	            csrfmiddlewaretoken: '{{ csrf_token }}'
	        },
	        contentType: "application/json;charset=utf-8",
	        dataType: "json",
	        success: function(data) {
				// pull the data out of the json we get back
	        	var html = '<h4><a href="/chado/feature/'+selection.id+'/">'+selection.name+'</a></h4>'
				// append the gene's families (if any)
				html += 'Family: ';
				if( selection.family != '' ) {
					html += '<a href="/chado/phylo/'+selection.family+'/">'+family_names[selection.family]+'</a>';
				} else {
					html += 'None';
				}
				html += '<br />';
	        	for (var i = 0; i < data.links.length; i++) {
	        	    for (var key in data.links[i]) {
	        	        html += '<a href="'+data.links[i][key]+'">'+key+'</a><br/>'
	        	    }
	        	}
	        	if (data.meta) {
	        	    html += '<p>'+data.meta+'</p>'
	        	}
	        	$('#toggle').html(html);
	        },
	        error: function(ts) { 
	        // this is for debugging, not production!
	            alert(ts.responseText);
	        }
	    });
		show_toggle();
	}

    // define the accessor used by the smith-waterman algorithm
    function get_family( item ) {
        return item.family;
    }

	// axis clicked callback function
	function axis_clicked( axis_selection, gene_selection, rail_selection ) {
		// add the track's links to the content box
		var html = '<h4><a href="/chado/organism/'+data.groups[axis_selection].species_id+'/">'+data.groups[axis_selection].species_name+'</a> - <a href="/chado/feature/'+data.groups[axis_selection].chromosome_id+'/">'+data.groups[axis_selection].chromosome_name+'</a></h4>';
		var l3 = '<a href="/chado/context_viewer/search/',
			l4 = '<a href="/chado/context_viewer/synteny/',
			l5 = '<a href="/chado/context_viewer/synteny2/';
		var genes = '<ul>';
		var families = [];
		gene_selection.each(function(d) {
			genes += '<li><a href="/chado/feature/'+d.id+'/">'+d.name+'</a>: '+d.fmin+' - '+d.fmax+'</li>';
			if( d.family != '' ) {
				genes += '<ul><li>Family: <a href="/chado/phylo/'+d.family+'/">'+family_names[d.family]+'</a></li></ul>'
			}
			// add links for the focus gene
			if( d.x == (track_length-1)/2 ) {
				l3 += d.id+'/';
				l4 += d.id+'/';
				l5 += d.id+'/';
			}
		});
		genes += '</ul>';
		l3 += '">Find similar tracks</a>';
		l4 += '">Synteny miner</a>';
		l5 += '">Synteny miner spaced</a>';
		html += l5+'<br />'+l4+'<br />'+l3+'<br />Genes:'+genes;
		$("#toggle").html(html);
		show_toggle();
	}
	
	// plot callback functions
	function brush_callback( selected_group ) {
        // make the sure the context data only has the query and selected group
        context_data.groups = [ context_data.groups[0], selected_group ];
        // perform the alignment
		var alignment = align( context_data.groups[0].genes, selected_group.genes, get_family );
        // clear the genes in the selected group
        context_data.groups[1].genes = [];
		// update the context data with the alignment
        var query_count = 0,
            query_length = context_data.groups[0].genes.length,
            pre_query = 0,
            insertion_count = 0;
        for( var i = 0; i < alignment[0].length; i++ ) {
            // keep track of how many selected genes come before the query genes
            if( alignment[0][i] == null && query_count == 0 ) {
                pre_query++;
            // it must be an insertion
            } else if( alignment[0][i] == null ) {
                // position the genes that come after the query genes
                if( query_count > query_length ) {
                    alignment[1][i].x = query_count++;
                    alignment[1][i].y = 1;
                    context_data.groups[1].genes.push(alignment[1][i]);
                // track how many genes were inserted
                } else {
                    insertion_count++;
                }
            // a deletion
            } else if( alignment[1][i] == null ) {
                query_count++;
            // an alignment
            } else {
                // position the genes that came before the query
                if( pre_query > 0 ) {
                    for( var j = 0; j < pre_query; j++ ) {
                        alignment[1][j].x = -1*(pre_query-(j+1));
                        alignment[1][j].y = 1;
                        context_data.groups[1].genes.push(alignment[1][j]);
                    }
                    pre_query = 0;
                // position the genes that go between query genes
                } else if( insertion_count > 0 ) {
                    var step = 1/(insertion_count+1);
                    for( var j = i-insertion_count; j < i; j++ ) {
                        alignment[1][j].x = query_count+(step*(i-j))-1;
                        alignment[1][j].y = 1;
                        context_data.groups[1].genes.push(alignment[1][j]);
                    }
                    insertion_count = 0;
                }
                alignment[1][i].x = query_count++;
                alignment[1][i].y = 1;
                context_data.groups[1].genes.push(alignment[1][i]);
            }
        }
        // plot the new context viewer
		context_viewer( "viewer",color, context_data, gene_clicked, axis_clicked );
	}
	
	// create the scatter plot matrix
	context_synteny( "synteny", color, data, gene_clicked, brush_callback );

    // the viewer only needs the query group
    context_data.groups = [context_data.groups[0]];
    // a helper function for sorting a group's genes
    function compare(a, b) {
        return a.fmin-b.fmin;
    }
    // sort the query group's genes
    context_data.groups[0].genes.sort(compare);
    // unpadte the corrdinates of the query group's genes
    for( var i = 0; i < context_data.groups[0].genes.length; i++ ) {
        context_data.groups[0].genes[i].y = 0;
        context_data.groups[0].genes[i].x = i;
    }
    // make the initial context viewer
    // add a flag that enables colors
	context_viewer( "viewer", color, context_data, gene_clicked, axis_clicked );

	// legend click callback
	function legend_click( family_selection, gene_selection ) {
		html = '<h4><a href="/chado/phylo/'+family_selection.id+'/">'+family_names[ family_selection ]+'</a></h4>';
		html += 'Genes:<ul>';
	    gene_selection.each(function(f) {
			html += '<li><a href="/chado/feature/'+f.id+'/">'+f.name+'</a>: '+f.fmin+' - '+f.fmax+'</li>';
		});
		html += '</ul>';
		$("#toggle").html(html);
		show_toggle();
	}

	// make the legend
	context_legend( 'legend', color, data, legend_click, false );
</script>
{% endblock %}
